File: app/src/main/java/me/fleey/futon/service/gateway/RateLimiter.kt
/*
 * Futon - Futon Daemon Client
 * Copyright (C) 2025 Fleey
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package me.fleey.futon.service.gateway

import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import org.koin.core.annotation.Single
import java.util.concurrent.ConcurrentHashMap

/**
 * Result of a rate limit check.
 */
sealed interface RateLimitResult {

  data object Allowed : RateLimitResult

  data class Limited(
    val retryAfterMs: Long,
    val currentCount: Int,
    val limit: Int,
  ) : RateLimitResult
}

/**
 * IP-based rate limiter using sliding window algorithm.
 */
interface RateLimiter {
  /** Set of currently rate-limited IPs (for monitoring) */
  val rateLimitedIps: StateFlow<Set<String>>

  /** Check if a request from the given IP is allowed */
  fun checkRequest(ip: String): RateLimitResult

  /** Record a request from the given IP */
  fun recordRequest(ip: String)

  fun clearIp(ip: String)

  fun clearAll()

  fun updateLimit(qps: Int)
}

@Single(binds = [RateLimiter::class])
class RateLimiterImpl(
  private val gatewayConfig: GatewayConfig,
) : RateLimiter {

  companion object {
    private const val TAG = "RateLimiter"
    private const val WINDOW_SIZE_MS = 1000L // 1 second window
    private const val CLEANUP_INTERVAL_MS = 60_000L // Cleanup every minute
  }

  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
  private val mutex = Mutex()

  // Map of IP -> list of request timestamps within the window
  private val requestWindows = ConcurrentHashMap<String, MutableList<Long>>()

  private val _rateLimitedIps = MutableStateFlow<Set<String>>(emptySet())
  override val rateLimitedIps: StateFlow<Set<String>> = _rateLimitedIps.asStateFlow()

  @Volatile
  private var currentLimit: Int = GatewayConfigData.DEFAULT_RATE_LIMIT_QPS

  init {
    scope.launch {
      gatewayConfig.config.collect { config ->
        currentLimit = config.rateLimitQps
      }
    }

    // Periodic cleanup of old entries
    scope.launch {
      while (true) {
        delay(CLEANUP_INTERVAL_MS)
        cleanup()
      }
    }
  }

  override fun checkRequest(ip: String): RateLimitResult {
    val now = System.currentTimeMillis()
    val windowStart = now - WINDOW_SIZE_MS

    val timestamps = requestWindows[ip] ?: return RateLimitResult.Allowed

    // Count requests within the window
    val recentRequests = synchronized(timestamps) {
      timestamps.removeAll { it < windowStart }
      timestamps.size
    }

    return if (recentRequests >= currentLimit) {
      // Calculate retry-after based on oldest request in window
      val oldestInWindow = synchronized(timestamps) {
        timestamps.minOrNull() ?: now
      }
      val retryAfter = (oldestInWindow + WINDOW_SIZE_MS) - now

      updateRateLimitedIps(ip, true)
      Log.w(TAG, "Rate limited IP: $ip (${recentRequests}/$currentLimit requests)")

      RateLimitResult.Limited(
        retryAfterMs = maxOf(retryAfter, 0),
        currentCount = recentRequests,
        limit = currentLimit,
      )
    } else {
      updateRateLimitedIps(ip, false)
      RateLimitResult.Allowed
    }
  }

  override fun recordRequest(ip: String) {
    val now = System.currentTimeMillis()
    val timestamps = requestWindows.computeIfAbsent(ip) { mutableListOf() }

    synchronized(timestamps) {
      timestamps.add(now)
      // Keep only recent timestamps to prevent memory growth
      val windowStart = now - WINDOW_SIZE_MS
      timestamps.removeAll { it < windowStart }
    }
  }

  override fun clearIp(ip: String) {
    requestWindows.remove(ip)
    updateRateLimitedIps(ip, false)
  }

  override fun clearAll() {
    requestWindows.clear()
    _rateLimitedIps.value = emptySet()
  }

  override fun updateLimit(qps: Int) {
    currentLimit = qps.coerceIn(
      GatewayConfigData.MIN_RATE_LIMIT,
      GatewayConfigData.MAX_RATE_LIMIT,
    )
  }

  private fun cleanup() {
    val now = System.currentTimeMillis()
    val windowStart = now - WINDOW_SIZE_MS

    val iterator = requestWindows.entries.iterator()
    while (iterator.hasNext()) {
      val entry = iterator.next()
      synchronized(entry.value) {
        entry.value.removeAll { it < windowStart }
        if (entry.value.isEmpty()) {
          iterator.remove()
        }
      }
    }

    // Update rate limited IPs set
    val stillLimited = requestWindows.entries
      .filter { (_, timestamps) ->
        synchronized(timestamps) { timestamps.size >= currentLimit }
      }
      .map { it.key }
      .toSet()

    _rateLimitedIps.value = stillLimited
  }

  private fun updateRateLimitedIps(ip: String, isLimited: Boolean) {
    val current = _rateLimitedIps.value
    _rateLimitedIps.value = if (isLimited) {
      current + ip
    } else {
      current - ip
    }
  }
}

File: daemon/core/auth/rate_limiter.cpp
/*
 * Futon - Android Automation Daemon
 * Copyright (C) 2025 Fleey
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "rate_limiter.h"
#include "core/error.h"

#include <algorithm>
#include <cmath>

namespace futon::core::auth {

    RateLimiter::RateLimiter(const RateLimitConfig &config)
            : config_(config) {
    }

    int64_t RateLimiter::current_time_ms() {
        auto now = std::chrono::steady_clock::now();
        return std::chrono::duration_cast<std::chrono::milliseconds>(
                now.time_since_epoch()
        ).count();
    }

    int64_t RateLimiter::calculate_backoff_ms(int failure_count) const {
        if (failure_count <= 0) return 0;

        // Exponential backoff: initial * multiplier^(failures-1)
        double backoff = config_.initial_backoff_ms *
                         std::pow(config_.backoff_multiplier, failure_count - 1);

        // Cap at max backoff
        return static_cast<int64_t>(std::min(backoff, static_cast<double>(config_.max_backoff_ms)));
    }

    RateLimitResult RateLimiter::check_allowed(uid_t uid) {
        std::lock_guard<std::mutex> lock(mutex_);

        int64_t now = current_time_ms();
        auto it = attempts_.find(uid);

        // No previous attempts - allowed
        if (it == attempts_.end()) {
            return {true, 0, config_.max_failures, nullptr};
        }

        AuthAttempt &attempt = it->second;

        // Check if reset window has passed (no attempts for a while)
        if (attempt.last_attempt_ms > 0 &&
            (now - attempt.last_attempt_ms) > config_.reset_window_ms) {
            attempt.reset();
            return {true, 0, config_.max_failures, nullptr};
        }

        // Check if currently locked out
        if (attempt.lockout_until_ms > now) {
            int64_t remaining = attempt.lockout_until_ms - now;
            return {
                    false,
                    remaining,
                    -1,
                    "Too many failed attempts. Please wait before retrying."
            };
        }

        // Check if max failures reached (but lockout expired)
        if (attempt.failed_count >= config_.max_failures) {
            // Lockout expired, but still at max failures - apply new lockout
            int64_t backoff = calculate_backoff_ms(attempt.failed_count);
            attempt.lockout_until_ms = now + backoff;

            return {
                    false,
                    backoff,
                    0,
                    "Maximum authentication attempts exceeded."
            };
        }

        // Allowed, but with reduced remaining attempts
        int remaining = config_.max_failures - attempt.failed_count;
        return {true, 0, remaining, nullptr};
    }

    void RateLimiter::record_success(uid_t uid) {
        std::lock_guard<std::mutex> lock(mutex_);

        auto it = attempts_.find(uid);
        if (it != attempts_.end()) {
            it->second.reset();
            FUTON_LOGI("Rate limiter: UID %d reset after successful auth", uid);
        }
    }

    void RateLimiter::record_failure(uid_t uid) {
        std::lock_guard<std::mutex> lock(mutex_);

        int64_t now = current_time_ms();
        AuthAttempt &attempt = attempts_[uid];

        // First failure in this window
        if (attempt.failed_count == 0) {
            attempt.first_failure_ms = now;
        }

        attempt.failed_count++;
        attempt.last_attempt_ms = now;

        // Calculate and apply lockout if threshold reached
        if (attempt.failed_count >= config_.max_failures) {
            int64_t backoff = calculate_backoff_ms(attempt.failed_count);
            attempt.lockout_until_ms = now + backoff;

            FUTON_LOGW("Rate limiter: UID %d locked out for %lld ms after %d failures",
                       uid, (long long) backoff, attempt.failed_count);
        } else {
            FUTON_LOGD("Rate limiter: UID %d failure %d/%d",
                       uid, attempt.failed_count, config_.max_failures);
        }
    }

    bool RateLimiter::is_locked_out(uid_t uid) const {
        std::lock_guard<std::mutex> lock(mutex_);

        auto it = attempts_.find(uid);
        if (it == attempts_.end()) return false;

        return it->second.lockout_until_ms > current_time_ms();
    }

    int RateLimiter::get_failed_count(uid_t uid) const {
        std::lock_guard<std::mutex> lock(mutex_);

        auto it = attempts_.find(uid);
        if (it == attempts_.end()) return 0;

        return it->second.failed_count;
    }

    int64_t RateLimiter::get_lockout_remaining_ms(uid_t uid) const {
        std::lock_guard<std::mutex> lock(mutex_);

        auto it = attempts_.find(uid);
        if (it == attempts_.end()) return 0;

        int64_t now = current_time_ms();
        if (it->second.lockout_until_ms <= now) return 0;

        return it->second.lockout_until_ms - now;
    }

    void RateLimiter::reset_uid(uid_t uid) {
        std::lock_guard<std::mutex> lock(mutex_);
        attempts_.erase(uid);
        FUTON_LOGI("Rate limiter: UID %d manually reset", uid);
    }

    void RateLimiter::reset_all() {
        std::lock_guard<std::mutex> lock(mutex_);
        attempts_.clear();
        FUTON_LOGI("Rate limiter: All UIDs reset");
    }

    void RateLimiter::cleanup_expired() {
        std::lock_guard<std::mutex> lock(mutex_);

        int64_t now = current_time_ms();
        int64_t cleanup_threshold = now - config_.reset_window_ms;

        for (auto it = attempts_.begin(); it != attempts_.end();) {
            // Remove entries that haven't had activity in reset_window_ms
            if (it->second.last_attempt_ms < cleanup_threshold) {
                FUTON_LOGD("Rate limiter: Cleaning up expired entry for UID %d", it->first);
                it = attempts_.erase(it);
            } else {
                ++it;
            }
        }
    }

    RateLimiter::Stats RateLimiter::get_stats() const {
        std::lock_guard<std::mutex> lock(mutex_);

        Stats stats{};
        stats.tracked_uids = attempts_.size();

        int64_t now = current_time_ms();
        for (const auto &[uid, attempt]: attempts_) {
            if (attempt.lockout_until_ms > now) {
                stats.locked_out_uids++;
            }
            stats.total_failures += attempt.failed_count;
        }

        return stats;
    }

} // namespace futon::core::auth
